### PR TITLE
[plugin-websocket] Add plugin to work with web sockets

### LINKED ISSUE #1402: Implement web sockets support
Our project aim to use Vividus for testing API (as well as GUI). Some endpoints use request-reply for communication between client and server, some of them use web sockets. The goal is to add web sockets support to Vividus.

Essentials:

- ability to fetch subscription messages continuously and verify them (while web socket is opened);
- ability to verify fields in response message;
- ability to close web socket connection and stop listening to new messages.

Some project specific details:

- GraphQL;
- Pattern we use as subscription URL: `ws://<serverAdress>/api/subscriptions?access_token=<token>`
- Example of subscription message structure: `{
  "data": {
    "<endpointName>": [
      {
        "<fieldName>": "<value>"
      }
         ]
  }
} 
`

### PR DESCRIPTION
**Pull request by uarlouski**
Closes #1402

### CODE REVIEW

File: docs/modules/plugins/pages/plugin-websocket.adoc [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+.Send text message
+[source,gherkin]
+----
+When I send text message `Hi everyone` over `chat` websocket
+----
+
+=== Disconnect
+
+Disconnects an existing websocket. All the received text messages are kept and can be drained into the variable using the drain step.
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-websocket.adoc:60]: `Disconnects from ...`

File: vividus-plugin-websocket/src/main/java/org/vividus/bdd/steps/websocket/WebSocketSteps.java [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+    private final TestContext testContext;
+    private final IBddVariableContext bddVariableContext;
+    private final SoftAssert softAssert;
+    private final JettyWebSocketClient webSocketClient;
+
+    public WebSocketSteps(Map<String, String> webSocketEndpoints, TestContext testContext,
+            IBddVariableContext bddVariableContext, SoftAssert softAssert)
+    {
+        this.webSocketEndpoints = new HashMap<>();
+        webSocketEndpoints.forEach((k, v) -> this.webSocketEndpoints.put(k, UriUtils.createUri(v)));
```
Comment (valfirst) [inline vividus-plugin-websocket/src/main/java/org/vividus/bdd/steps/websocket/WebSocketSteps.java:67]: don't we have automatic mapping to `Map<String, URI>` ?

File: vividus-plugin-websocket/src/main/java/org/vividus/bdd/steps/websocket/WebSocketSteps.java [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+    @When("I wait with `$timeout` timeout until count of text messages received over `$webSocketConnectionKey`"
+            + " websocket is $comparisonRule `$expectedCount`")
+    public void waitForTextMessages(Duration timeout, String webSocketConnectionKey, ComparisonRule comparisonRule,
+            int expectedCount)
+    {
+        WebSocket webSocket = getWebSocket(webSocketConnectionKey, false);
+        Matcher<Integer> countMatcher = comparisonRule.getComparisonRule(expectedCount);
+        Integer result = new DurationBasedWaiter(timeout, Duration.ofSeconds(1))
+                .wait(() -> webSocket.getMessages().size(), countMatcher::matches);
+        softAssert.assertThat("Total count of received over WebSocket messages", result, countMatcher);
```
Comment (valfirst) [inline vividus-plugin-websocket/src/main/java/org/vividus/bdd/steps/websocket/WebSocketSteps.java:163]: `"Total count of messages received over WebSocket"`

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-websocket/build.gradle [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+    implementation project(':vividus-util')
+
+    implementation group: 'org.springframework', name: 'spring-websocket', version: '5.3.4'
+    implementation group: 'org.eclipse.jetty.websocket', name: 'websocket-client', version: '9.4.36.v20210114'
+
+    testImplementation platform(group: 'org.junit', name: 'junit-bom', version: versions.junit)
+    testImplementation(group: 'org.junit.jupiter', name: 'junit-jupiter')
+    testImplementation(group: 'org.mockito', name: 'mockito-junit-jupiter', version: versions.mockito)
+    testImplementation(group: 'com.github.valfirst', name: 'slf4j-test', version: versions.slf4jTest)
+}
```
Comment (valfirst) [inline vividus-plugin-websocket/build.gradle:15]: Also please disable publishing till next release

**Review by valfirst — APPROVED** [review]

File: vividus-plugin-websocket/build.gradle [hunk 5]
```diff
@@ -0,0 +1,17 @@
+project.description = 'Vividus plugin for WebSockets'
+
+dependencies {
+    api project(':vividus-bdd-engine')
+    implementation project(':vividus-soft-assert')
+    implementation project(':vividus-util')
+
+    implementation(group: 'org.springframework', name: 'spring-websocket', version: versions.spring)
+    implementation(group: 'org.eclipse.jetty.websocket', name: 'websocket-client', version: '9.4.36.v20210114')
```
Comment (ikalinin1) [inline vividus-plugin-websocket/build.gradle:9]: 9.4.38.v20210224 ? :)

**Review by ikalinin1 — APPROVED** [review]

**Pull request merged**